### PR TITLE
IMEX improvements

### PR DESCRIPTION
**Pull request by mt5555**
Several improvements for the NH IMEX code:

1. better splitting between explicit and implicit/stiff parts
2. new 2nd order IMEX algorithm with final implicit stage - much better results with topography.  tstep_type=10.  
3. Rewrite Jacobian in terms of d(phi)/deta, to increase precision and simplify formulas
4. Includes @ambrad 's 2nd order accurate PPM vertical remap branch that was needed for testing
5. disable dp and exner scaling of potential temperature hyperviscosity (this was causing stability problems)

[nonBFB] for homme suite (all theta tests) and eam_theta suite.

### CODE REVIEW

File: components/cam/bld/configure [hunk 1]
```diff
@@ -1843,7 +1843,7 @@ if ($offline_dyn) { $cfg_cppdefs .= ' -DOFFLINE_DYN'; }
 # Spectral Element dycore
 if($dyn_pkg eq 'se'){
     if ($spmd eq 'ON'){	$cfg_cppdefs .=" -D_MPI"; }
-    $cfg_cppdefs .= " -DCAM  -D_WK_GRAD -D_PRIM ";
+    $cfg_cppdefs .= " -DCAM  -DENERGY_DIAGNOSTICS -D_WK_GRAD -D_PRIM ";
```
Comment (mt5555) [inline components/cam/bld/configure]: this should be removed

File: components/homme/src/theta-l/eos.F90 [hunk 2]
```diff
@@ -275,64 +323,72 @@ subroutine get_dirk_jacobian(JacL,JacD,JacU,dt2,dp3d,phi_i,pnh,exact,&
...
-          JacU(k-1,:,:) = (g*dt2)**2 * (dpnh_dp_i(:,:,k-1)-dpnh_dp_i_epsie(:,:,k-1))/epsie
+           JacU(k-1,:,:) = delta_mu(:,:,k-1)
         else
-          JacL(k,:,:)   = (g*dt2)**2 * (dpnh_dp_i(:,:,k+1)-dpnh_dp_i_epsie(:,:,k+1))/epsie
-          JacD(k,:,:)   = 1 + (g*dt2)**2 * (dpnh_dp_i(:,:,k)-dpnh_dp_i_epsie(:,:,k))/epsie
-          JacU(k-1,:,:) = (g*dt2)**2 * (dpnh_dp_i(:,:,k-1)-dpnh_dp_i_epsie(:,:,k-1))/epsie
+           JacL(k,:,:)   = delta_mu(:,:,k+1)
+           JacU(k-1,:,:) = delta_mu(:,:,k-1)
         end if
+
```
Comment (oksanaguba) [inline components/homme/src/theta-l/eos.F90]: Above, for inexact jacobian:
k==1: JacL(1) is set
k==nlev: JacU(71) is set
2<k<nlev: JacL(2:71), JacU(1:70) are set
so, all indices are accounted for, but it is not readable. delta_mu is precomputed, so lines above could be simplified.

**Comment by oksanaguba** [comment]
On anvil against today's next next+this branch eam_theta suite: all fails as DIFFS, except for SL that did not run. ERS tests are PASS for layout restarts and DIFF for baselines. HOMME basoSL test did not run (even though i did resoft with some generic file from anvil webpage).

**Comment by oksanaguba** [comment]
Ran homme test thru standalone cmake, not cime, and SL runs again. Theta tests diff:
```Test project /home/onguba/as/runhomme/bld-integrator/test_execs
      Start  1: verifyBaselineResults
 1/21 Test  #1: verifyBaselineResults ............   Passed    0.63 sec
      Start  2: swtc1
 2/21 Test  #2: swtc1 ............................   Passed   23.42 sec
      Start  3: swtc2
 3/21 Test  #3: swtc2 ............................   Passed    8.10 sec
      Start  4: swtc5
 4/21 Test  #4: swtc5 ............................   Passed   35.13 sec
      Start  5: swtc6
 5/21 Test  #5: swtc6 ............................   Passed    2.77 sec
      Start  6: baro2b
 6/21 Test  #6: baro2b ...........................   Passed  122.29 sec
      Start  7: baro2c
 7/21 Test  #7: baro2c ...........................   Passed   15.48 sec
      Start  8: baro2d
 8/21 Test  #8: baro2d ...........................   Passed  170.16 sec
      Start  9: baroCamMoist
 9/21 Test  #9: baroCamMoist .....................   Passed   42.97 sec
      Start 10: baroCamMoistSL
10/21 Test #10: baroCamMoistSL ...................   Passed    1.35 sec
      Start 11: baroCamMoist-acc
11/21 Test #11: baroCamMoist-acc .................   Passed    7.88 sec
      Start 12: thetah-test22
12/21 Test #12: thetah-test22 ....................***Failed   31.97 sec
Submitting 1 jobs
Running test thetah-test22-run ... /home/onguba/as/runhomme/bld-integrator/tests/thetah-test22/thetah-test22-run.sh > thetah-test22-run.out 2> thetah-test22-run.err
test thetah-test22-run was run successfully
Test name = thetah-test22
Examining cprnc reference comparison output files
Test thetah-test22: no netcdf CHECKREF output files, skipping CHECKREF test.
############################################################################
  The diff using CPRNC has passed
############################################################################
Examining cprnc baseline comparison output files
file = dcmip2012_test2_21.nc
The files are different: DIFF_RESULT=DIFFERENT
############################################################################
CPRNC returned the following RMS differences
 RMS ps                               1.3442E-02            NORMALIZED  1.3462E-07
 RMS T                                1.0678E-04            NORMALIZED  3.5774E-07
 RMS u                                1.5951E-04            NORMALIZED  4.4675E-06
 RMS v                                1.2686E-04            NORMALIZED  1.0013E-02
 RMS omega                            5.6051E-03            NORMALIZED  1.4309E-02
############################################################################

      Start 13: thetanh-test22
13/21 Test #13: thetanh-test22 ...................***Failed   35.58 sec
Submitting 1 jobs
Running test thetanh-test22-run ... /home/onguba/as/runhomme/bld-integrator/tests/thetanh-test22/thetanh-test22-run.sh > thetanh-test22-run.out 2> thetanh-test22-run.err
test thetanh-test22-run was run successfully
Test name = thetanh-test22
Examining cprnc reference comparison output files
Test thetanh-test22: no netcdf CHECKREF output files, skipping CHECKREF test.
############################################################################
  The diff using CPRNC has passed
############################################################################
Examining cprnc baseline comparison output files
file = dcmip2012_test2_21.nc
The files are different: DIFF_RESULT=DIFFERENT
############################################################################
CPRNC returned the following RMS differences
 RMS ps                               9.2282E-02            NORMALIZED  9.2420E-07
 RMS T                                2.3881E-04            NORMALIZED  8.0011E-07
 RMS u                                2.4494E-05            NORMALIZED  6.8602E-07
 RMS v                                2.6704E-05            NORMALIZED  4.0590E-03
 RMS omega                            1.5411E-03            NORMALIZED  1.0072E-02
############################################################################

      Start 14: preqx-TC-ftype4
14/21 Test #14: preqx-TC-ftype4 ..................   Passed    3.15 sec
      Start 15: thetah-TC-ftype4
15/21 Test #15: thetah-TC-ftype4 .................***Failed    3.55 sec
Submitting 1 jobs
Running test thetah-TC-ftype4-run ... /home/onguba/as/runhomme/bld-integrator/tests/thetah-TC-ftype4/thetah-TC-ftype4-run.sh > thetah-TC-ftype4-run.out 2> thetah-TC-ftype4-run.err
test thetah-TC-ftype4-run was run successfully
Test name = thetah-TC-ftype4
Examining cprnc reference comparison output files
Test thetah-TC-ftype4: no netcdf CHECKREF output files, skipping CHECKREF test.
############################################################################
  The diff using CPRNC has passed
############################################################################
Examining cprnc baseline comparison output files
file = dcmip2016_test21.nc
The files are different: DIFF_RESULT=DIFFERENT
############################################################################
CPRNC returned the following RMS differences
 RMS ps                               1.1898E-01            NORMALIZED  1.1722E-06
 RMS T                                7.1950E-04            NORMALIZED  2.9355E-06
 RMS w                                6.0862E-06            NORMALIZED  4.1235E-02
 RMS Q                                9.0669E-07            NORMALIZED  1.7713E-04
 RMS Q4                               3.6893E-08            NORMALIZED  3.6893E-08
############################################################################

      Start 16: thetah-TC
16/21 Test #16: thetah-TC ........................***Failed    3.52 sec
Submitting 1 jobs
Running test thetah-TC-run ... /home/onguba/as/runhomme/bld-integrator/tests/thetah-TC/thetah-TC-run.sh > thetah-TC-run.out 2> thetah-TC-run.err
test thetah-TC-run was run successfully
Test name = thetah-TC
Examining cprnc reference comparison output files
Test thetah-TC: no netcdf CHECKREF output files, skipping CHECKREF test.
############################################################################
  The diff using CPRNC has passed
############################################################################
Examining cprnc baseline comparison output files
file = dcmip2016_test21.nc
The files are different: DIFF_RESULT=DIFFERENT
############################################################################
CPRNC returned the following RMS differences
 RMS ps                               1.0731E-01            NORMALIZED  1.0572E-06
 RMS T                                7.0832E-04            NORMALIZED  2.8899E-06
 RMS w                                5.9607E-06            NORMALIZED  4.0513E-02
 RMS Q                                8.2391E-07            NORMALIZED  1.6096E-04
 RMS Q4                               1.5678E-14            NORMALIZED  1.5678E-14
############################################################################

      Start 17: thetanh-TC
17/21 Test #17: thetanh-TC .......................***Failed    7.59 sec
Submitting 1 jobs
Running test thetanh-TC-run ... /home/onguba/as/runhomme/bld-integrator/tests/thetanh-TC/thetanh-TC-run.sh > thetanh-TC-run.out 2> thetanh-TC-run.err
test thetanh-TC-run was run successfully
Test name = thetanh-TC
Examining cprnc reference comparison output files
Test thetanh-TC: no netcdf CHECKREF output files, skipping CHECKREF test.
############################################################################
  The diff using CPRNC has passed
############################################################################
Examining cprnc baseline comparison output files
file = dcmip2016_test21.nc
The files are different: DIFF_RESULT=DIFFERENT
############################################################################
CPRNC returned the following RMS differences
 RMS ps                               1.1094E-01            NORMALIZED  1.0930E-06
 RMS T                                6.7673E-04            NORMALIZED  2.7611E-06
 RMS w                                1.5722E-06            NORMALIZED  6.1711E-02
 RMS Q                                8.7317E-07            NORMALIZED  1.7058E-04
 RMS Q4                               1.1118E-14            NORMALIZED  1.1118E-14
############################################################################

      Start 18: thetanh-TC-nudiv
18/21 Test #18: thetanh-TC-nudiv .................***Failed    8.64 sec
Submitting 1 jobs
Running test thetanh-TC-nudiv-run ... /home/onguba/as/runhomme/bld-integrator/tests/thetanh-TC-nudiv/thetanh-TC-nudiv-run.sh > thetanh-TC-nudiv-run.out 2> thetanh-TC-nudiv-run.err
test thetanh-TC-nudiv-run was run successfully
Test name = thetanh-TC-nudiv
Examining cprnc reference comparison output files
Test thetanh-TC-nudiv: no netcdf CHECKREF output files, skipping CHECKREF test.
############################################################################
  The diff using CPRNC has passed
############################################################################
Examining cprnc baseline comparison output files
file = dcmip2016_test21.nc
The files are different: DIFF_RESULT=DIFFERENT
############################################################################
CPRNC returned the following RMS differences
 RMS ps                               1.1809E-01            NORMALIZED  1.1634E-06
 RMS T                                7.0219E-04            NORMALIZED  2.8650E-06
 RMS w                                1.5106E-06            NORMALIZED  7.9813E-02
 RMS Q                                8.9999E-07            NORMALIZED  1.7583E-04
 RMS Q4                               1.1415E-14            NORMALIZED  1.1415E-14
############################################################################

      Start 19: thetanh-TC-nutop
19/21 Test #19: thetanh-TC-nutop .................***Failed    7.94 sec
Submitting 1 jobs
Running test thetanh-TC-nutop-run ... /home/onguba/as/runhomme/bld-integrator/tests/thetanh-TC-nutop/thetanh-TC-nutop-run.sh > thetanh-TC-nutop-run.out 2> thetanh-TC-nutop-run.err
test thetanh-TC-nutop-run was run successfully
Test name = thetanh-TC-nutop
Examining cprnc reference comparison output files
Test thetanh-TC-nutop: no netcdf CHECKREF output files, skipping CHECKREF test.
############################################################################
  The diff using CPRNC has passed
############################################################################
Examining cprnc baseline comparison output files
file = dcmip2016_test21.nc
The files are different: DIFF_RESULT=DIFFERENT
############################################################################
CPRNC returned the following RMS differences
 RMS ps                               1.1091E-01            NORMALIZED  1.0927E-06
 RMS T                                6.7444E-04            NORMALIZED  2.7517E-06
 RMS w                                1.5520E-06            NORMALIZED  6.6461E-02
 RMS Q                                8.6988E-07            NORMALIZED  1.6994E-04
 RMS Q4                               4.5055E-08            NORMALIZED  4.5055E-08
############################################################################

      Start 20: thetanhwet-TC
20/21 Test #20: thetanhwet-TC ....................***Failed    8.87 sec
Submitting 1 jobs
Running test thetanhwet-TC-run ... /home/onguba/as/runhomme/bld-integrator/tests/thetanhwet-TC/thetanhwet-TC-run.sh > thetanhwet-TC-run.out 2> thetanhwet-TC-run.err
test thetanhwet-TC-run was run successfully
Test name = thetanhwet-TC
Examining cprnc reference comparison output files
Test thetanhwet-TC: no netcdf CHECKREF output files, skipping CHECKREF test.
############################################################################
  The diff using CPRNC has passed
############################################################################
Examining cprnc baseline comparison output files
file = dcmip2016_test21.nc
The files are different: DIFF_RESULT=DIFFERENT
############################################################################
CPRNC returned the following RMS differences
 RMS ps                               1.2419E-01            NORMALIZED  1.2236E-06
 RMS T                                6.7854E-04            NORMALIZED  2.7684E-06
 RMS w                                1.6368E-06            NORMALIZED  6.9804E-02
 RMS Q                                9.7985E-07            NORMALIZED  1.9142E-04
 RMS Q2                               3.1495E-08            NORMALIZED  1.9919E+00
############################################################################

      Start 21: templates
21/21 Test #21: templates ........................   Passed    2.42 sec

62% tests passed, 8 tests failed out of 21

Total Test time (real) = 543.45 sec

The following tests FAILED:
	 12 - thetah-test22 (Failed)
	 13 - thetanh-test22 (Failed)
	 15 - thetah-TC-ftype4 (Failed)
	 16 - thetah-TC (Failed)
	 17 - thetanh-TC (Failed)
	 18 - thetanh-TC-nudiv (Failed)
	 19 - thetanh-TC-nutop (Failed)
	 20 - thetanhwet-TC (Failed)```

**Comment by oksanaguba** [comment]
Ran these tests but there may be 1 more commit coming in, so, there may be another round of tests. Not clear if all dcmip tests were looked at (if their plots look as expected), but that is out of scope.

**Comment by oksanaguba** [comment]
Merged to next.

**Pull request merged**